### PR TITLE
Don't apply `require_partition_filter` to a temporary table

### LINKED ISSUE #64: `require_partition_filter=true` with the `insert_overwrite` strategy doesn't work
### Describe the bug
When running an incremental model with the `insert_overwrite` strategy and the `require_partition_filter=true` option, it seems that the `require_partition_filter` option is applied to the temporary table.

```
  Query error: Cannot query over table 'sandbox-project.logs.event__dbt_tmp' without a filter over column(s) 'event_time' that can be used for partition elimination at [75:46]
```

### Steps To Reproduce
1. Create an incremental model with `insert_overwrite`
```sql
# test_insert_overwrite.sql
{{
  config(
    materialized="incremental",
    incremental_strategy='insert_overwrite',
    require_partition_filter=true,
    alias="test_insert_overwrite",
    partition_by={
      "field": "event_time",
      "data_type": "timestamp",
      "granularity": "day",
    },
  )
}}

SELECT
  CURRENT_TIMESTAMP() AS event_time
```

2. Execute `dbt run -s test_insert_overwrite` twice
```
Completed with 1 error and 0 warnings:

Database Error in model test_insert_overwrite (models/test_insert_overwrite.sql)
  Query error: Cannot query over table 'sandbox-project.jaffle_shop.test_insert_overwrite__dbt_tmp' without a filter over column(s) 'event_time' that can be used for partition elimination at [35:46]
  compiled SQL at target/run/jaffle_shop/models/test_insert_overwrite.sql
```

### Screenshots and log output
The query below was generated when the second time execution of the incremental model. As we can see, the query to create the temporary table `test_insert_overwrite__dbt_tmp` includes `require_partition_filter=true`, because the macro to generate is `create_table_as`. The `create_table_as` macro automatically applies all the same options to even the temporary table. 

- https://github.com/dbt-labs/dbt-bigquery/blob/main/dbt/include/bigquery/macros/materializations/incremental.sql#L71

```sql
      -- generated script to merge partitions into `sandbox-project`.`jaffle_shop`.`test_insert_overwrite`
      declare dbt_partitions_for_replacement array<timestamp>;
      declare _dbt_max_partition timestamp default (
          select max(event_time) from `sandbox-project`.`jaffle_shop`.`test_insert_overwrite`
          where event_time is not null
      );

      
      
        -- 1. create a temp table
        

  create or replace table `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
  partition by timestamp_trunc(event_time, day)
  
  OPTIONS(
      expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 12 hour),
    
      require_partition_filter=True
    )
  as (
    

SELECT
  CURRENT_TIMESTAMP() AS event_time
  );
      

      -- 2. define partitions to update
      set (dbt_partitions_for_replacement) = (
          select as struct
              array_agg(distinct timestamp_trunc(event_time, day))
          from `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
      );

      
      -- 3. run the merge statement
      

    merge into `sandbox-project`.`jaffle_shop`.`test_insert_overwrite` as DBT_INTERNAL_DEST
        using (
        select * from `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
      ) as DBT_INTERNAL_SOURCE
        on FALSE

    when not matched by source
         and timestamp_trunc(DBT_INTERNAL_DEST.event_time, day) in unnest(dbt_partitions_for_replacement) 
        then delete

    when not matched then insert
        (`event_time`)
    values
        (`event_time`)

;

      -- 4. clean up the temp table
      drop table if exists `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
```

### System information
**The output of `dbt --version`:**
```
installed version: 0.21.0
   latest version: 0.21.0

Up to date!

Plugins:
  - bigquery: 0.21.0
  - snowflake: 0.21.0
  - redshift: 0.21.0
  - postgres: 0.21.0
```

**The operating system you're using:**
- Mac OS 11.6 (20G165)

**The output of `python --version`:**
- Python 3.8.8

### PR DESCRIPTION
**Pull request by yu-iskw**
resolves #64

### Description
We won't apply the `require_partition_filter` and `partition_expiration_days` options to a temporary table, because the incremental materialization with the `insert_overwrite` strategy applies the options even to the temporary table to over-write partitions.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-bigquery next" section.

### CODE REVIEW

**Comment by yu-iskw** [comment]
## Testing execution with the developmed package

```bash
$ dbt run -s test_insert_overwrite
10:11:26 | [ info  ] | Running with dbt=1.0.0-rc1
10:11:26 | [ warn  ] | * Deprecation Warning: The `source-paths` config has been deprecated in favor of
`model-paths`. Please update your `dbt_project.yml` configuration to reflect
this change.
10:11:26 | [ warn  ] | * Deprecation Warning: The `source-paths` config has been deprecated in favor of
`model-paths`. Please update your `dbt_project.yml` configuration to reflect
this change.
10:11:26 | [ info  ] | Found 9 models, 20 tests, 0 snapshots, 0 analyses, 188 macros, 0 operations, 3 seed files, 0 sources, 0 exposures, 0 metrics
10:11:26 | [ info  ] |
10:11:26 | [ info  ] | Concurrency: 5 threads (target='dev')
10:11:26 | [ info  ] |
10:11:26 | [ info  ] | 1 of 1 START incremental model jaffle_shop.test_insert_overwrite................ [RUN]
10:11:26 | [ info  ] | 1 of 1 OK created incremental model jaffle_shop.test_insert_overwrite........... [SCRIPT (24.0 Bytes processed) in 7.29s]
10:11:26 | [ info  ] |
10:11:26 | [ info  ] | Finished running 1 incremental model in 13.29s.
10:11:26 | [ info  ] |
10:11:26 | [ info  ] | Completed successfully
10:11:26 | [ info  ] |
Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

**Comment by yu-iskw** [comment]
## The run query
The `CREATE` statement for the temporary table surely doesn't contain the `require_partition_filter` and `partition_expiration_days` options.

```sql

      -- generated script to merge partitions into `sandbox-project`.`jaffle_shop`.`test_insert_overwrite`
      declare dbt_partitions_for_replacement array<timestamp>;
  
        -- 1. create a temp table
        

  create or replace table `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
  partition by timestamp_trunc(event_time, day)
  
  OPTIONS(
      expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 12 hour)
    )
  as (
    

SELECT
  CURRENT_TIMESTAMP() AS event_time
  );
      

      -- 2. define partitions to update
      set (dbt_partitions_for_replacement) = (
          select as struct
              array_agg(distinct timestamp_trunc(event_time, day))
          from `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
      );

      
      -- 3. run the merge statement
      

    merge into `sandbox-project`.`jaffle_shop`.`test_insert_overwrite` as DBT_INTERNAL_DEST
        using (
        select * from `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
      ) as DBT_INTERNAL_SOURCE
        on FALSE

    when not matched by source
         and timestamp_trunc(DBT_INTERNAL_DEST.event_time, day) in unnest(dbt_partitions_for_replacement) 
        then delete

    when not matched then insert
        (`event_time`)
    values
        (`event_time`)

;

      -- 4. clean up the temp table
      drop table if exists `sandbox-project`.`jaffle_shop`.`test_insert_overwrite__dbt_tmp`
```

**Comment by yu-iskw** [comment]
@jtcohen6 Can you take a look? I would like to fix the problem before releasing dbt 1.0.0.

File: dbt/adapters/bigquery/impl.py [hunk 1]
```diff
@@ -756,20 +756,22 @@ def get_table_options(
...
             opts['kms_key_name'] = "'{}'".format(config.get('kms_key_name'))
 
-        if config.get('require_partition_filter'):
-            opts['require_partition_filter'] = config.get(
-                'require_partition_filter')
-
-        if config.get('partition_expiration_days') is not None:
-            opts['partition_expiration_days'] = config.get(
-                'partition_expiration_days')
+        if temporary:
```
Comment (jtcohen6) [inline dbt/adapters/bigquery/impl.py:762]: Do we need all of this logic here, vs. the simpler logic proposed by @prratek in https://github.com/dbt-labs/dbt-core/pull/3386:
```python
        if config.get('require_partition_filter') and not temporary:
            opts['require_partition_filter'] = config.get(
                'require_partition_filter')
```

**Review by jtcohen6 — COMMENTED** [review]
Thanks for the quick PR @yu-iskw! I agree, I'd love to have this simple change in before v1. If possible, there are a few related pieces:
- To get this working for `merge` strategy as well, we need to add `(DBT_INTERNAL_DEST.{{ partition_by.field }} is not null or DBT_INTERNAL_DEST.{{ partition_by.field }} is null)` to the set of predicates passed in by the "merge" conditional branch of the BigQuery incremental materialization [here](https://github.com/dbt-labs/dbt-bigquery/blob/95ca84b5100c2b940751d8a65aaeb914df211ac7/dbt/include/bigquery/macros/materializations/incremental.sql#L115-L128), to be used by [the `get_merge_sql` macro](https://github.com/dbt-labs/dbt-core/blob/3d28b6704ce53446293fc0acecb0b57cd0f6f6ba/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L24)
- We'll want some integration test cases, I think one of each incremental strategy + `require_partition_filter`, to ensure we don't regress from this behavior. That's where we got tripped up in https://github.com/dbt-labs/dbt-core/pull/3386, because the test case for `datetime`-type partitioning + `require_partition_filter` wouldn't pass due to (I believe) a BigQuery bug. We can leave out or comment out that test case accordingly.

I won't consider those related pieces to be true blockers, since just this change alone is an improvement over status quo, but I would love a more complete solution to #24 if possible.

**Comment by yu-iskw** [comment]
@jtcohen6 Thank you for sharing the context. I agree with ideally solving a more complete solution with one-shot. Let me look into the merge strategy. If I come up with any idea shortly, I will combine them all to the pull request. If not, I would like to merge the simpler solution for now.

File: dbt/adapters/bigquery/impl.py (hunk 1 above)
Comment (yu-iskw) [reply dbt/adapters/bigquery/impl.py:762]: It is ok with me. The reason why I made the nested conditions under `if temporary:` is the two conditions are a set for a temporary table. I will unnest the two.

**Comment by jtcohen6** [comment]
@yu-iskw If you're able to get this PR merge-ready, I'd be ok with moving ahead with the narrower fix for now

**Comment by yu-iskw** [comment]
@jtcohen6 Sounds good. I will address #24 in another pull request. Let's merge it.

**Review by jtcohen6 — APPROVED** [review]

**Pull request merged**